File: Main/rumour/rumour.py
import random

# receiver class
class Receiver:
    def __init__(self, name, status, message):
        self.name = name
        self.status = status
        self.message = message

class Sender:
    def __init__(self, name):
        self.name = name

    def feedback(self, node, message):
        # check if node has the message if it has, 
        # use p
        # robability to decide whether to gossip or not
        if node.status == 1:
            p = 3
            probability = random.choices([0, 1],((p - 1) / p, 1 / p),k=1)
            if probability[0] == 0:
                node.status = 1
                node.message = message
        # if a node has not received the message
        #  do not stop the gossip
        else:
            node.status = 1
            node.message = message
        print(node.name, node.status, node.message)

    def blind(self, node, message):
        # calculate probability
        p = 3
        determinant = 1 / p
        probability = random.choices([1 / p, ((p - 1) / p)], k=1)
        # if probability is not 1/p continue gossipping
        if determinant != probability[0]:
            node.status = 1
            node.message = message
        print(node.name, node.status, node.message)

    # fixed probability
    def fixed(self, nodes, message):
        global node
        k = 3
        count = 0
        for node in nodes:
            if node.status == 1:
                count += 1
        # if the k nodes have not received the message
        #  gossip to them
        if count != k:
            for node in nodes:
                node.status = 1
                node.message = message
            print(node.name, node.status, node.message)


A = Receiver('A', 0, "No message")
B = Receiver('B', 1, "No message")
C = Receiver('C', 0, "No message")
D = Receiver('D', 1, "No message")
nodes = [A, B, C, D]


s = Sender('S')
s.fixed(nodes, "Fixed Probability")
s.blind(D, "Blind probability")
s.feedback(B, "Feedback Probability")

File: Main/rumour/rumour.cpp
#include <stdio.h>
#include <iostream>
#include <ctime>
using namespace std;

class Receivers{
	int status;
	char name;
	public:
	void UpdateStatus();
	void ShowStatus();
	void setName(char nm);
	int getStatus();
	Receivers(){
		status = 0;
	}
};

class Sender{
	int k, counter;
	public:
	void Blind(Receivers *obj, int len);
	void Feedback(Receivers *obj, int len);
	void Fixed(Receivers *obj,int len);

	Sender(){
	k=3;
	counter =0;
	}

};
void Sender::Blind(Receivers *obj,int len){

    srand((unsigned) time(0));
     int prob = (rand()%10)+1;
// Only looses interest if probability prob is less than  k
//It does not check how many nodes have received the Gossip so no counter is kept
while(k<=prob){
	//If the probability is out of range it sends the message whether or not the node had received it earlier
	//Select A Random Receiver
	int ran = ((rand()%len)+1)-1;
	cout <<"\n Receiver Node: "<<ran;
	cout<<"\n Status of Receiver: "<<obj[ran].getStatus();
	obj[ran].UpdateStatus();
	prob = (rand()%10)+1;

}
//End Blind
}


void Sender::Feedback(Receivers *obj,int len){

    srand((unsigned) time(0));
     int prob = (rand()%10)+1;
// Only Stops Sending Gossip if probability prob is less than  k

    while(k<=prob){
    //If the probability is out of range it sends the message whether or not the node had received it earlier

     //Select A Random Receiver
    int ran = ((rand()%len)+1)-1;
   //Check receiver Status
    int stat = obj[ran].getStatus();
    counter = counter +1;



    cout <<"\n Receiver Node: "<<ran;
    cout<<"\n Status of Receiver: "<<obj[ran].getStatus();
    obj[ran].UpdateStatus();
    prob = (rand()%10)+1;

    }
//End Feedback
}

void Sender::Fixed(Receivers *obj, int len){
    //Send Message
    srand((unsigned) time(0));
//Loose interest once k nodes have confirmed receiving the message
    while(counter<k){

        int ran = ((rand()%len)+1)-1;
     int stat = obj[ran].getStatus();

     if (stat==1){
        //The Receiver Selected Got The message
        //Increase Counter
         counter = counter + 1;
        cout<<"\n Counter1: "<<counter;
        continue;
     }else{
        //The Receiver Did Not Get The Gossip
        //Pass The Gossip by changing status value
        obj[ran].UpdateStatus();
       // counter = counter + 1;
        //cout<<"\n Counter else: "<<counter;


     }


     //cout <<"Random: "<<ran;
    }



}
void Receivers::UpdateStatus(){
    status = 1;
}
void Receivers::setName(char nm){
    name = nm;
}
void Receivers::ShowStatus(){
     cout<<"Status: "<<status;
}
int Receivers::getStatus(){
     return status;
}

int main(){
    Receivers obj[4];
    Sender P;
    char name;
    //Size of object
     int si = sizeof(obj)/sizeof(*obj);
    for(int i =0; i<=si;i++){
        cout<<"\n Enter Node Name: ";
        cin>>name;
        obj[i].setName(name);
    }


}

